codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java

```diff
@@ -32,89 +32,81 @@
+    private static final byte READ_LITERAL_HEADER_NAME_LENGTH = 5;
+    private static final byte READ_LITERAL_HEADER_NAME = 6;
+    private static final byte READ_LITERAL_HEADER_VALUE_LENGTH_PREFIX = 7;
+    private static final byte READ_LITERAL_HEADER_VALUE_LENGTH = 8;
+    private static final byte READ_LITERAL_HEADER_VALUE = 9;
 
     private final DynamicTable dynamicTable;
     private final HuffmanDecoder huffmanDecoder;
-    private final int maxHeaderSize;
+    private final int maxHeadersLength;
     private int maxDynamicTableSize;
     private int encoderMaxDynamicTableSize;
     private boolean maxDynamicTableSizeChangeRequired;
 
-    private long headerSize;
-    private State state;
-    private IndexType indexType;
-    private int index;
-    private boolean huffmanEncoded;
-    private int skipLength;
-    private int nameLength;
-    private int valueLength;
-    private CharSequence name;
-
-    private enum State {
-        READ_HEADER_REPRESENTATION,
-        READ_MAX_DYNAMIC_TABLE_SIZE,
-        READ_INDEXED_HEADER,
-        READ_INDEXED_HEADER_NAME,
-        READ_LITERAL_HEADER_NAME_LENGTH_PREFIX,
-        READ_LITERAL_HEADER_NAME_LENGTH,
-        READ_LITERAL_HEADER_NAME,
-        SKIP_LITERAL_HEADER_NAME,
-        READ_LITERAL_HEADER_VALUE_LENGTH_PREFIX,
-        READ_LITERAL_HEADER_VALUE_LENGTH,
-        READ_LITERAL_HEADER_VALUE,
-        SKIP_LITERAL_HEADER_VALUE
-    }
-
     /**
-     * Creates a new decoder.
+     * Create a new instance.
+     * @param maxHeadersLength The maximum size (in bytes) that is allowed for a single header decode operation.
+     * @param maxHeaderTableSize
+     * <a href="https://tools.ietf.org/html/rfc7540#section-6.5.2">SETTINGS_HEADER_TABLE_SIZE</a>.
+     * @param initialHuffmanDecodeCapacity The initial size of the byte array used to do huffman decoding.
      */
-    public Decoder(int maxHeaderSize, int maxHeaderTableSize, int initialHuffmanDecodeCapacity) {
+    public Decoder(int maxHeadersLength, int maxHeaderTableSize, int initialHuffmanDecodeCapacity) {
         dynamicTable = new DynamicTable(maxHeaderTableSize);
-        this.maxHeaderSize = maxHeaderSize;
+        this.maxHeadersLength = maxHeadersLength;
         maxDynamicTableSize = maxHeaderTableSize;
         encoderMaxDynamicTableSize = maxHeaderTableSize;
         maxDynamicTableSizeChangeRequired = false;
         huffmanDecoder = new HuffmanDecoder(initialHuffmanDecodeCapacity);
-        reset();
-    }
-
-    private void reset() {
-        headerSize = 0;
-        state = State.READ_HEADER_REPRESENTATION;
-        indexType = IndexType.NONE;
     }
 
     /**
      * Decode the header block into header fields.
+     * <p>
+     * This method assumes the entire header block is contained in {@code in}.
      */
-    public void decode(ByteBuf in, Http2Headers headers) throws IOException {
+    public void decode(ByteBuf in, Http2Headers headers) throws IOException, Http2Exception {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this method to throw at most one checked exception instead of: java.io.IOException, io.netty.handler.codec.http2.Http2Exception [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1160)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "decode" is 49 which is greater than 25 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)